src/Host/Client/Impl/Host/RHostConnector.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.WebSockets.Client;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.Logging;
+using Microsoft.R.Host.Protocol;
+using Newtonsoft.Json;
+
+namespace Microsoft.R.Host.Client.Host {
+    internal abstract class RHostConnector : IRHostConnector {
+        private static readonly TimeSpan HeartbeatTimeout =
+#if DEBUG
+            // In debug mode, increase the timeout significantly, so that when the host is paused in debugger,
+            // the client won't immediately timeout and disconnect.
+            TimeSpan.FromMinutes(10);
+#else
+            TimeSpan.FromSeconds(5);
+#endif
+
+        private readonly LinesLog _log;
+        private HttpClient _broker;
+        private string _interpreterId;
+
+        protected HttpClient Broker => _broker;
+
+        public bool IsDisposed { get; private set; }
+
+        protected RHostConnector(string interpreterId) {
+            _interpreterId = interpreterId;
+            _log = new LinesLog(FileLogWriter.InTempFolder("Microsoft.R.Host.BrokerConnector"));
+        }
+
+        protected void CreateHttpClient() {
+            _broker = new HttpClient(GetHttpClientHandler()) {
+                Timeout = TimeSpan.FromSeconds(30)
+            };
+
+            _broker.DefaultRequestHeaders.Accept.Clear();
+            _broker.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        protected abstract HttpClientHandler GetHttpClientHandler();
+
+        protected abstract void ConfigureWebSocketRequest(HttpWebRequest request);
+
+        public virtual void Dispose() {
+            IsDisposed = true;
```

Comment:
There are 2 classes that can be used here - `DisposeToken` and `DisposableBag`. Both are thread-safe.